### ISSUE TITLE
inject does not properly resolve the result of a cartesian-product

### ISSUE THREAD

**Issue by jmchuster** [geofffranks/spruce#169]
Notice that inject does properly insert the result of a concat, but I'm guessing cartesian-product is along a different code path because it's actually defining a new array.

http://play.spruce.cf/#a5ca315a4212beaa21f7c0a6c538a001

file:

``` yaml
meta:
  fast-tests:
    - isolation
  slow-tests:
    - integration
  master-fast-tests:
    jobs:
      (( cartesian-product "master-" meta.fast-tests "-tests" ))
  master-slow-tests:
    jobs:
    # (( cartesian-product "master-" meta.fast-tests "-tests" ))
      - (( concat "master-" meta.slow-tests.0 "-tests" ))

groups:
  - name: master
    .: (( inject meta.master-fast-tests ))
    ..: (( inject meta.master-slow-tests ))
    jobs:
    - master-dependencies-test
```

expected output:

``` yaml
groups:
- jobs:
  - master-isolation-tests
  - master-integration-tests
  - master-dependencies-test
  name: master
```

actual output:

``` yaml
groups:
- jobs:
  - master-integration-tests
  - master-dependencies-test
  name: master
```

**Comment by thomasmmitchell** [comment]
Whats happening is that inject is cowardly refusing to overwrite the existing list with a string. The same behavior can be exhibited here: http://play.spruce.cf/#85b09c5afc3537a1fc9a6e5dbc09c004

Cartesian product just happens to be in a map, while concat is in a list. This could be fixed by moving the evaluation of inject into eval phase (because right now, inject is invariably evaluated before concat or cartesian), but that would cause a whole slew of other problems.

**Comment by jhunt** [comment]
I don't think we can move inject into eval phase, since it is specifically designed to be used at merge time, file-at-a-time.

**Comment by thomasmmitchell** [comment]
Another option is to have duplicate hash values such as this be put into a temporarily value and marked as such somehow, and if at the end of eval phase it has been reconciled back into a list (as it would in this case), it is appended... or merged... or whatever inject does... and if it hasn't become a list, it is tossed away as it would be otherwise. This gets to keep the order separation of inject while also allowing the order to not screw things up as much.

Its messy, but its the first thing that comes to mind

**Comment by geofffranks** [comment]
What about somehow specifying an option to cartesian product to allow it to insert itself into the array its defined in (if it's defined inside an array), rather than stick the array at the current element? Then you could make the cartesian product operator string a member of the array it's trying to be merged with, and it doesn't get overwritten.

We could do this via an env variable, or flag to spruce to control the cartesian product behavior on a global level, or add a new inline-cartesian-product operator so that the behavior is more granular.

**Comment by thomasmmitchell** [comment]
I would say add it as `(( inline-cartesian-product ... ))`. It seems like a niche thing

**Comment by jmchuster** [comment]
So, it doesn't seem like `inject` is the right tool for the job here, since there is a specific requirement around its ordering.

What about using `grab`?  Does the following make more sense?

```yaml
meta:
  fast-tests:
    - isolation
  slow-tests:
    - integration
  master-fast-tests:
    (( cartesian-product "master-" meta.fast-tests "-tests" ))
  master-slow-tests:
    - (( concat "master-" meta.slow-tests.0 "-tests" ))

groups:
- name: master
  jobs: (( grab
             meta.master-fast-tests
             meta.master-slow-tests
             "master-dependencies-test"
             "master-docker-build" ))
```

http://play.spruce.cf/#e9dfbe4879abf0c46ae4dad00bb4edd3

**Comment by geofffranks** [comment]
According to @thomasmmitchell this was fixed in #220
